src/envoy/http/authn/filter_context.cc

```diff
@@ -74,7 +74,42 @@ void FilterContext::setPrincipal(const iaapi::PrincipalBinding& binding) {
 
 bool FilterContext::getJwtPayload(const std::string& issuer,
                                   std::string* payload) const {
-  const auto filter_it =
+  return getJwtPayloadFromEnvoyJwtFilter(issuer, payload) ||
+         getJwtPayloadFromIstioJwtFilter(issuer, payload);
+}
+
+bool FilterContext::getJwtPayloadFromEnvoyJwtFilter(const std::string& issuer,
+                                  std::string* payload) const {
+  // Try getting the Jwt payload from Envoy jwt_authn filter.
+  auto filter_it =
+      dynamic_metadata_.filter_metadata().find(
+        Extensions::HttpFilters::HttpFilterNames::get().JwtAuthn);
+  if (filter_it == dynamic_metadata_.filter_metadata().end()) {
+    ENVOY_LOG(debug, "No dynamic_metadata found for filter {}",
+              Extensions::HttpFilters::HttpFilterNames::get().JwtAuthn);
+    return false;
+  }
+
+  const auto& data_struct = filter_it->second;
+
+  const auto entry_it = data_struct.fields().find(issuer);
+  if (entry_it == data_struct.fields().end()) {
+    return false;
+  }
+
+  if (entry_it->second.struct_value().fields().empty()) {
+    return false;
+  }
+
+  // Serialize the payload from Envoy jwt filter first before writing it to |payload|.
+  Protobuf::util::MessageToJsonString(entry_it->second.struct_value(), payload);
```

Comment:
I think we don't need to convert it back to string once the istio jwt filter is completely removed, which should improve the performance and also simplify the code. Could you put a TODO as a reminder?

Reply:
It depends whether we want to use a structured proto, i.e payload, or the proto struct directly in the next step. Either way, I agree this can be revised later.

Reply:
Didn't see @diemtvu's comment, so I added a TODO anyway. I think we can revise later.